src/map.cpp

```diff
@@ -6910,6 +6910,108 @@ void map::restock_fruits( const tripoint &p, int time_since_last_actualize )
     }
 }
 
+void map::produce_sap( const tripoint &p, int time_since_last_actualize )
+{
+    if( time_since_last_actualize == 0 ) {
+        return;
+    }
+
+    const ter_id tid = ter( p );
+    if( tid != t_tree_maple_tapped ) {
+        return;
+    }
+
+    // How many turns to produce 1 charge (250 ml) of sap?
+    int turns_season = DAYS( calendar::season_length() );
+    int producing_length = int( 0.75f * turns_season );
+    int turns_to_produce = producing_length / ( MAPLE_SAP_PER_SEASON * 4 );
+
+    // How long of this time_since_last_actualize have we been in the producing period (late winter, early spring)?
+    int time_producing = 0;
+
+    if( time_since_last_actualize >= calendar::year_turns() ) {
+        time_producing = producing_length;
+    } else {
+        // We are only procuding sap on the intersection with the sap producing season.
+        int early_spring_end = int( 0.5f * turns_season );
+        int late_winter_start = int( 3.75f * turns_season );
+
+        calendar last_actualize = calendar::turn - time_since_last_actualize;
+        int last_actualize_tof = last_actualize.turn_of_year();
+        bool last_producing = (
+            last_actualize_tof >= late_winter_start ||
+            last_actualize_tof < early_spring_end
+        );
+        int current_tof = calendar::turn.turn_of_year();
+        bool current_producing = (
+            current_tof >= late_winter_start ||
+            current_tof < early_spring_end
+        );
+
+        int non_procuding_length = int( 3.25f * turns_season );
+
+        if( last_producing && current_producing ) {
+            if( time_since_last_actualize < non_procuding_length ) {
+                time_producing = time_since_last_actualize;
+            } else {
+                time_producing = time_since_last_actualize - non_procuding_length;
+            }
+        } else if ( !last_producing && !current_producing ) {
+            if( time_since_last_actualize > non_procuding_length ) {
+                time_producing = time_since_last_actualize - non_procuding_length;
+            }
+        } else if ( last_producing && !current_producing ) {
+            // We hit the end of early spring
+            if( last_actualize_tof < early_spring_end ) {
+                time_producing = early_spring_end - last_actualize_tof;
+            } else {
+                time_producing = calendar::year_turns() - last_actualize_tof + early_spring_end;
+            }
+        } else if ( !last_producing && current_producing ) {
+            // We hit the start of late winter
+            if( current_tof >= late_winter_start ) {
+                time_producing = current_tof - late_winter_start;
+            } else {
+                time_producing = int( 0.25f * turns_season ) + current_tof;
+            }
+        }
+    }
+
+    // Not enough time to produce 1 charge of sap
+    if( time_producing < turns_to_produce ) {
```

Comment:
I'm not sure how I would use `roll_remainder` in that comparison. I am not rounding a double.

In order to avoid less sap being produced for a player checking the area often, what about adding a random fraction of turns_to_produce to time_producing, something like `time_producing += rng( 0, int( 0.3f * turns_to_produce ) );` before the comparison? What do you think?